### PR TITLE
Rewrite battle prediction in JS

### PR DESCRIPTION
**Pull request by Adrymne**
As noted in `BattlePrediction.js`, run `grunt concat:battlePredictionDev` to get it to work on a dev build.

### CODE REVIEW

**Comment by sinsinpub** [comment]
besides, i wanna request new features to bp modules:
 * As a new attribute of the result ship object, add 'damage' to it to record the damage made by each ship of both player and enemy fleet. This can be used to predict MVP ship.
 * Furthermore, to get rid of the `beginHPs`, `endHPs` things, damage gauge rate (useful for field `allyNoDamage`) and the rank predication can be also put into the module.

as an implementation reference, see [poi lib-battle](https://github.com/poooi/lib-battle/blob/master/simulator.es).

**Comment by Adrymne** [comment]
I appreciate the help - I didn't really do a good job when updating the code in `engage()` and engageNight()`. I'm mostly done with a fix for #2183, so I'll probably finish that off before I work on the new bits.

> A question, at night battle, items of damecon for player combined fleet should be 12-array or 6 only for escort? old helper needs to put 6 of escort into 12-array.

That's a bug, it should be 12. The main fleet is still set up, it just isn't modified.

File: src/library/modules/BattlePrediction/rank/Rank.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    return Object.keys(battleResult).reduce((result, key) => {
+      return Object.assign(result, {
+        [key]: hideOverkill(omitEmptySlots(battleResult[key])),
+      });
+    }, {});
+  };
+
+  const predictors = {
+    ld_airbattle: 'airRaid',
+  };
```
Comment (sinsinpub) [inline src/library/modules/BattlePrediction/rank/Rank.js]: oh, should not be full string matched here... the `api_name` is not official api property, it's added in our `Kcsapi.js` (shorten ver of request url), that's why uses `indexOf` matching.

File: src/library/modules/BattlePrediction/rank/Battle.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      battle: { isPlayerNoDamage, isEnemyFlagshipSunk, isPlayerFlagshipTaiha },
+    } = KC3BattlePrediction.rank;
+
+    const sunk = getSunkCount(resultFleets);
+    const ships = getShipCount(resultFleets);
+    if (sunk.player === 0) {
+      if (sunk.enemy === ships.enemy) {
+        return isPlayerNoDamage(initialFleets, resultFleets) ? 'SS' : 'S';
+      }
+      if (ships.enemy > 1 && sunk.enemy >= Math.floor(ships.enemy * (2 / 3))) {
```
Comment (sinsinpub) [inline src/library/modules/BattlePrediction/rank/Battle.js]: wondering where is `2/3` from? EO's memo using `0.7` and they do not full equal for combined fleet:
```
> [1,2,3,4,5,6,7,8,9,10,11,12].map(v => Math.floor(v*(2/3)))
[0, 1, 2, 2, 3, 4, 4, 5, 6, 6, 7, 8]
> [1,2,3,4,5,6,7,8,9,10,11,12].map(v => Math.floor(v*0.7))
[0, 1, 2, 2, 3, 4, 4, 5, 6, 7, 7, 8]
```

Comment (Adrymne) [reply src/library/modules/BattlePrediction/rank/Battle.js]: I don't remember where exactly I pulled it from, I was looking at a bunch of sources. I'll change it to match EO.

Honestly, I'm not particularly happy with how that function turned out, I might go back and rewrite it later.

**Comment by Adrymne** [comment]
Unsurprisingly, implementing MVP prediction involved a fairly substantial rewrite. I didn't implement a UI for it - I'll leave that to someone more familiar with that part of the code.

Once it's been checked, I'll tidy things up and prepare it for merge.

**Comment by sinsinpub** [comment]
Add more commits on `dev-bp2` to check if MVP prediction is capable, current rules:
 * battle rank D or E (using predicated rank) Incapable. Later ship sometimes gets MVP, likely flagship lower priority for these rank.
 * starting from night battle, no aerial combat or no bombing damage is capable.
 * sum total bombing damage, check if there is a ship carrying bomber, her predicted dealt damage + total bombing damage is > damage dealt by MVP ship, Incapable if true.
 * otherwise is capable.
might add some rules about combined fleet with night battle.

**Pull request merged**